<a href="https://colab.research.google.com/github/koyomin9zx/CS313.K21-Homework/blob/master/homework2_cau1%2B2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
import pandas as pd
from collections import defaultdict

name=[]
region=[]
dict_plant=defaultdict(list)
with open('plants.data', encoding='ISO-8859-1') as f:
    lines=f.readlines()
    
for line in lines:
    line=line.strip('\n')
    if line.split(",")[0] not in name:
      name.append(line.split(",")[0])

    region_split=line.split(",")[1:]
    dict_plant[name[-1]]=region_split
    for r in region_split:
        if r not in region:
            region.append(r)

In [33]:
df=pd.DataFrame(columns=['names']+region)
for key, values in df.iteritems():
  df[key]=[0] * len(name)
df['names']=name

df

names  fl  nc  ct  dc  ...  de  mo  nj  dengl  gl
0                               abelia   0   0   0   0  ...   0   0   0      0   0
1                 abelia x grandiflora   0   0   0   0  ...   0   0   0      0   0
2                          abelmoschus   0   0   0   0  ...   0   0   0      0   0
3               abelmoschus esculentus   0   0   0   0  ...   0   0   0      0   0
4                abelmoschus moschatus   0   0   0   0  ...   0   0   0      0   0
...                                ...  ..  ..  ..  ..  ...  ..  ..  ..    ...  ..
34776  zuckia brandegeei var. plummeri   0   0   0   0  ...   0   0   0      0   0
34777                      zygophlebia   0   0   0   0  ...   0   0   0      0   0
34778           zygophlebia sectifrons   0   0   0   0  ...   0   0   0      0   0
34779                      zygophyllum   0   0   0   0  ...   0   0   0      0   0
34780               zygophyllum fabago   0   0   0   0  ...   0   0   0      0   0

[34781 rows x 71 columns]

In [0]:
i=1
for key, values in dict_plant.items():
  for value in values:
    df.loc[df['names'] == key, value] = 1 
  print("----processed " +str(i)+" samples---\n")
  i+=1

df.to_csv('plants.csv', sep=',',index=None)

In [0]:
import numpy as np
freq=[]
for key, values in df.iteritems():
  count_plant_per_region = df.pivot_table(index=[key], aggfunc='size')
  print(key + " : "+str(count_plant_per_region[1]))
  freq.append(count_plant_per_region[1])

cal=np.array(freq)
cal=np.delete(cal,0)

In [0]:
#find min
cal.min()

#find max
cal.max()

#sum
cal.sum()

Link download tệp [plants.csv](https://drive.google.com/file/d/1WpWhqPjAGjIf3ZSBG59plJ04qI3sZEu2/view?usp=sharing) 

a. Có tất cả bao nhiêu loài cây : **34781 loài**

b. Có tất cả bao nhiêu vùng phân bố : **70 vùng**

c. Số loài cây trên mỗi vùng phân bố: 

    fl : 6621
    nc : 5926
    ct : 4391
    dc : 3080
    hi : 3804
    il : 5167
    ky : 4555
    la : 5154
    md : 5108
    mi : 4734
    ms : 4815
    sc : 5432
    va : 5638
    pr : 4781
    vi : 2185
    ak : 2969
    az : 6778
    ca : 11676
    co : 5465
    ga : 5942
    id : 5129
    in : 4440
    ia : 3652
    me : 3969
    ma : 4963
    mn : 3929
    mt : 4800
    nv : 5670
    nh : 3635
    nm : 6403
    ny : 5773
    oh : 4772
    or : 7028
    pa : 5474
    ri : 3295
    tn : 4900
    ut : 6041
    vt : 3713
    wa : 5654
    wv : 4062
    wi : 4321
    wy : 4710
    ab : 3408
    bc : 4875
    lb : 1433
    mb : 3023
    nb : 2856
    nf : 2188
    nt : 2024
    ns : 2844
    nu : 979
    on : 5068
    pe : 1841
    qc : 4272
    sk : 2846
    yt : 2100
    fraspm : 1210
    ks : 3869
    ne : 3281
    nd : 2682
    ok : 4651
    sd : 3185
    tx : 8483
    al : 5702
    ar : 4610
    de : 3630
    mo : 4638
    nj : 4822
    dengl : 479
    gl : 561

d. Vùng phân bố có ít loài cây nhất, cho biết số lượng và tỉ lệ % :

    Vùng phân bố ít nhất: dengl - 479 loài
    tỉ lệ %: 1.38 %

e. Vùng phân bố có nhiều loài cây nhất, cho biết số lượng và tỉ lệ % :

    Vùng phân bố nhiều nhất: ca - 11676 loài
    tỉ lệ %: 33.57 %

f. Trung bình một vùng phân bố có bao nhiêu loài cây : 

    Trung bình 1 vùng có: 4315,91 loài cây
